tests/test_jetmet_tools.py

```diff
@@ -252,27 +343,113 @@ def test_jet_transformer():
         assert('pt_'+unc+'_up' in met.columns)
         assert('phi_'+unc+'_up' in met.columns)
 
-def test_jet_correction_uncertainty_sources():
-    from coffea.jetmet_tools import JetCorrectionUncertainty
-
-    counts, test_eta, test_pt = dummy_jagged_eta_pt()
 
-    junc_names = []
-    levels = []
-    for name in dir(evaluator):
-        if 'Summer16_23Sep2016V3_MC_UncertaintySources_AK4PFPuppi' in name:
-            junc_names.append(name)
-            levels.append(name.split('_')[-1])
-        #test for underscore in dataera
-        if 'Fall17_17Nov2017_V6_MC_UncertaintySources_AK4PFchs_AbsoluteFlavMap' in name:
-            junc_names.append(name)
-            levels.append(name.split('_')[-1])
-    junc = JetCorrectionUncertainty(**{name: evaluator[name] for name in junc_names})
-
-    print(junc)
-
-    juncs = junc.getUncertainty(JetEta=test_eta, JetPt=test_pt)
-
-    for level, corrs in juncs:
-        assert(level in levels)
-        assert(corrs.shape[0] == test_eta.shape[0])
+def test_corrected_jets_factory():
+    import os
+    from coffea.jetmet_tools import CorrectedJetsFactory, CorrectedMETFactory, JECStack
+
+    events = None
+    cache = {}
+    from coffea.nanoevents import NanoEventsFactory
+    factory = NanoEventsFactory.from_file(os.path.abspath('tests/samples/nano_dy.root'))
+    events = factory.events()
+    
+    jec_stack_names = ['Summer16_23Sep2016V3_MC_L1FastJet_AK4PFPuppi',
+                       'Summer16_23Sep2016V3_MC_L2Relative_AK4PFPuppi',
+                       'Summer16_23Sep2016V3_MC_L2L3Residual_AK4PFPuppi',
+                       'Summer16_23Sep2016V3_MC_L3Absolute_AK4PFPuppi',
+                       'Spring16_25nsV10_MC_PtResolution_AK4PFPuppi',
+                       'Spring16_25nsV10_MC_SF_AK4PFPuppi']
+    for key in evaluator.keys():
+        if 'Summer16_23Sep2016V3_MC_UncertaintySources_AK4PFPuppi' in key:
+            jec_stack_names.append(key)
+
+    jec_inputs = {name: evaluator[name] for name in jec_stack_names}
+    jec_stack = JECStack(jec_inputs)
+
+    name_map = jec_stack.blank_name_map
+    name_map['JetPt'] = 'pt'
+    name_map['JetMass'] = 'mass'
+    name_map['JetEta'] = 'eta'
+    name_map['JetA'] = 'area'
+    
+    jets = events.Jet
+    
+    jets['pt_raw'] = jets['rawFactor'] * jets['pt']
+    jets['mass_raw'] = jets['rawFactor'] * jets['mass']
```

Comment:
Shouldn't it be `(1 - jets['rawFactor'])` instead?
I'm getting incorrect results otherwise.

Reply:
Yes, it should

Reply:
My bad. Can you post a PR and I'll get it integrated, do a backport for the tests.

Reply:
https://github.com/CoffeaTeam/coffea/pull/387